raiden/network/proxies/token_network.py

```diff
@@ -814,94 +800,104 @@ def _set_total_deposit(
             self.client.poll(transaction_hash)
             receipt_or_none = check_transaction_threw(self.client, transaction_hash)
 
-        transaction_executed = gas_limit is not None
-        if not transaction_executed or receipt_or_none:
-            if transaction_executed:
-                block = receipt_or_none["blockNumber"]
-            else:
-                block = checking_block
+            if receipt_or_none:
+                # Because the gas estimation succeeded it is known that:
+                # - The channel was open.
+                # - The account had enough balance to deposit
+                # - The account had enough balance to pay for the gas (however
+                #   there is a race condition for multiple transactions #3890)
 
-            self.proxy.jsonrpc_client.check_for_insufficient_eth(
-                transaction_name="setTotalDeposit",
-                transaction_executed=transaction_executed,
-                required_gas=self.gas_measurements["TokenNetwork.setTotalDeposit"],
-                block_identifier=block,
-            )
-            error_type, msg = self._check_why_deposit_failed(
-                channel_identifier=channel_identifier,
-                partner=partner,
-                token=token,
-                amount_to_deposit=amount_to_deposit,
-                total_deposit=total_deposit,
-                transaction_executed=transaction_executed,
-                block_identifier=block,
-            )
+                if receipt_or_none["cumulativeGasUsed"] == gas_limit:
```

Comment:
There are many checks missing from this branch:

- Was the state of the [channel open](https://github.com/raiden-network/raiden-contracts/blob/8ff92c6ce4d23537b703d8f98f5ef2e8678b3add/raiden_contracts/data/source/raiden/TokenNetwork.sol#L320)?
- Did an [deposit overflow happen](https://github.com/raiden-network/raiden-contracts/blob/8ff92c6ce4d23537b703d8f98f5ef2e8678b3add/raiden_contracts/data/source/raiden/TokenNetwork.sol#L356)? Highly unlikely, but should be checked anyways (useful with testnet deployments when huge values are used)
- [the participants maximum deposit amount](https://github.com/raiden-network/raiden-contracts/blob/8ff92c6ce4d23537b703d8f98f5ef2e8678b3add/raiden_contracts/data/source/raiden/TokenNetwork.sol#L324), another transaction could have deposited on the channel, again a critical section bug or another account making a deposit.
- [the network maximum deposit](https://github.com/raiden-network/raiden-contracts/blob/8ff92c6ce4d23537b703d8f98f5ef2e8678b3add/raiden_contracts/data/source/raiden/TokenNetwork.sol#L348), this is a race condition that can happen when another deposit happen.
- The approved amount, did it change meanwhile? (token.approve could have been called concurrently if we had a badly defined critical section)
- Does the user have enough tokens? (token.transfer could have been called meanwhile, again with a ill define critical section, which btw is the case ATM)

Reply:
> Does the user have enough tokens? (token.transfer could have been called meanwhile, again with a ill define critical section, which btw is the case ATM)

All done except of this. I think if the transaction fails because of a lack of "approved" funds, then the `allowance` is what should be checked and it's done there. However, if the lack of tokens was the reason, it would be the call to `approve` to fail not the total deposit transaction.